firebase-firestore/ktx/src/main/kotlin/com/google/firebase/firestore/ktx/serialization/NestedMapEncoder.kt

```diff
@@ -0,0 +1,213 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.ktx.serialization
+
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.descriptors.StructureKind
+import kotlinx.serialization.descriptors.elementNames
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.encoding.CompositeEncoder
+import kotlinx.serialization.modules.EmptySerializersModule
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.serializer
+
+/**
+ * The entry point of Firestore Kotlin Serialization Process. It encodes a custom @Serializable
+ * Kotlin object into a nested map of primitives. SDK users don’t need to directly interact with
+ * [NestedMapEncoder].
+ *
+ * <p>For a custom @Serializable object, at compile time, a serializer will be generated by the
+ * Kotlin serialization compiler plugin (or a custom serializer can be manually passed in). The
+ * structure information of the custom @Serializable object will be recorded inside of the
+ * serializer’s descriptor (i.e. the name/type of each property to be encoded, the annotation on
+ * each property).
+ *
+ * <p>Based on the descriptor’s information, during the encoding process, a nested map will be
+ * generated. Each property which has its own structure (i.e. a nested custom object, a nested list)
+ * will be encoded as an embedded map nested inside. At the end, the nested map at depth == 1 will
+ * be returned as the nested map representing the entire custom @Serializable object.
+ *
+ * @param map the nested map that records the encoded result of the original @Serializable object.
+ * @param depth the current encoding depth inside of the nested map.
+ * @param propertyNamesToBeEncoded a list of all property names to be encoded at the current depth.
+ */
+class NestedMapEncoder(
```

Comment:
Since you asked the other day about alternative names, maybe just `MapEncoder` ? I think it's implicit that is going to end up being nested.

Reply:
Alternatively `FirestoreMapEncoder` could work, as it will understand firestore-specific annotations

Reply:
Thank you for the suggestions. I think I can call the Java interface `MapEncoder`, the Kotlin subclass `KtxMapEncoder`, and this kotlin subclass will have two fields, `FirestoreMapEncoder` and `FirestoreMapDecoder`, probably another `config_setting_field` as well, so user can decide they want snake_case, CAPITAL, or lowerCamelCase for their enum fields (my stretch goal), and two methods need to overwrite: encodeToMap, and decodeToMap (convert between Object and Map)